File: app/src/main/java/com/arnichem/arnichem_barcode/TransactionsView/validate_dc/SecondValidateDcActivity.java
package com.arnichem.arnichem_barcode.TransactionsView.validate_dc;

import static com.arnichem.arnichem_barcode.Reset.APIClient.delivery_validation;

import android.Manifest;
import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.location.Location;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.app.ActivityCompat;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.android.volley.AuthFailureError;
import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.toolbox.StringRequest;
import com.arnichem.arnichem_barcode.Barcode.LaserScannerActivity;
import com.arnichem.arnichem_barcode.Barcode.NewScanner;
import com.arnichem.arnichem_barcode.OnItemClickListener;
import com.arnichem.arnichem_barcode.PrintReceipt.DeliveryPrint.ViewDeliveryPrint;
import com.arnichem.arnichem_barcode.PrintReceipt.MainPrintActivity;
import com.arnichem.arnichem_barcode.R;
import com.arnichem.arnichem_barcode.digital_signature.ActivityDigitalSignature;
import com.arnichem.arnichem_barcode.util.SharedPref;
import com.arnichem.arnichem_barcode.view.VolleySingleton;
import com.example.easywaylocation.EasyWayLocation;
import com.example.easywaylocation.Listener;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationServices;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import java.io.File;
import java.io.IOException;
import java.lang.reflect.Type;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.Base64;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// Placeholder for CylinderData (replace with actual implementation)

@RequiresApi(api = Build.VERSION_CODES.O)
public class SecondValidateDcActivity extends AppCompatActivity implements OnItemClickListener , Listener {
    String latitude = "0", logitude = "0", address = "0";

            private RecyclerView recyclerView;
            private TextView dcNoTextView;
            private TextView invoiceNoTextView;
            private TextView customerNameTextView;
    private EasyWayLocation easyWayLocation;

    private EditText newScan;

    private Button submitButton;
            private Button uploadSign;
            private ConstraintLayout constraintSigned;
            private ImageView signedImg;
    private String inputHolder = "";

    private ImageView closeImg;
            private ProgressDialog dialog;
            private List<CylinderData> cylinderList = new ArrayList<>();
            private String digitalSignPath = "";
            private String digitalSignBase64 = "";
            private static final String TAG = "SecondValidateDcActivity";
            private boolean isAllFabsVisible = false;
            private FusedLocationProviderClient fusedLocationClient;

            @Override
            protected void onCreate(Bundle savedInstanceState) {
                super.onCreate(savedInstanceState);
                setContentView(R.layout.activity_second_validate_dc);

                getSupportActionBar().setDisplayHomeAsUpEnabled(true);
                getSupportActionBar().setTitle("Validate DC");

                setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
                LocalBroadcastManager.getInstance(this).registerReceiver(mServiceReceiver,
                        new IntentFilter("digital_sign"));
                easyWayLocation = new EasyWayLocation(this, false,true,this);


                // Initialize location client
                fusedLocationClient = LocationServices.getFusedLocationProviderClient(this);

                dcNoTextView = findViewById(R.id.dcNoTextView);
                invoiceNoTextView = findViewById(R.id.invoiceNoTextView);
                customerNameTextView = findViewById(R.id.customerNameTextView);
                recyclerView = findViewById(R.id.recyclerView);
                submitButton = findViewById(R.id.submitButton);
                uploadSign = findViewById(R.id.uploadSign);
                constraintSigned = findViewById(R.id.constraintSigned);
                signedImg = findViewById(R.id.signedImg);
                closeImg = findViewById(R.id.closeImg);
                newScan = findViewById(R.id.newScan);
                
                // Fix: Use OnKeyListener instead of dispatchKeyEvent
                newScan.setOnKeyListener(new View.OnKeyListener() {
                    @Override
                    public boolean onKey(View v, int keyCode, KeyEvent event) {
                        if ((keyCode == KeyEvent.KEYCODE_ENTER || keyCode == KeyEvent.KEYCODE_NUMPAD_ENTER)
                                && event.getAction() == KeyEvent.ACTION_UP) {
                            
                            String text = newScan.getText().toString().trim();
                            if (!text.isEmpty()) {
                                Log.d(TAG, "Validate Scan Detected: " + text);
                                validateCylinder(text);
                            }
                            newScan.setText("");
                            newScan.requestFocus();
                            return true;
                        }
                        return false;
                    }
                });

                // Focus Protection
                newScan.setOnFocusChangeListener(new View.OnFocusChangeListener() {
                    @Override
                    public void onFocusChange(View v, boolean hasFocus) {
                        if (!hasFocus) {
                             newScan.postDelayed(() -> newScan.requestFocus(), 50);
                        }
                    }
                });



                Intent intent = getIntent();
                String dcno = intent.getStringExtra("dcno");
                String invoiceNo = intent.getStringExtra("invoiceNo");
                String customerName = intent.getStringExtra("customerName");
                String cylinderNumbersJson = intent.getStringExtra("cylinderNumbers");

                Type listType = new TypeToken<List<CylinderData>>() {}.getType();
                cylinderList = new Gson().fromJson(cylinderNumbersJson, listType);
                if (cylinderList == null) cylinderList = new ArrayList<>();

                dcNoTextView.setText("Dc No: " + dcno);
                invoiceNoTextView.setText("Invoice No: " + invoiceNo);
                customerNameTextView.setText("Customer Name: " + customerName);

                recyclerView.setLayoutManager(new LinearLayoutManager(this));
                recyclerView.setAdapter(new CylinderAdapter(this, cylinderList, this::validateCylinder));


                submitButton.setOnClickListener(v -> {
                    dialog = new ProgressDialog(this);
                    dialog.setTitle("Submitting");
                    dialog.setMessage("Please wait...");
                    dialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
                    dialog.setCancelable(false);
                    dialog.show();
                    postUsingVolley();
                });

                uploadSign.setOnClickListener(v -> {
                    startActivity(new Intent(this, ActivityDigitalSignature.class).putExtra("type", "validate"));
                });

                closeImg.setOnClickListener(v -> {
                    SharedPref.getInstance(this).setSign("");
                    constraintSigned.setVisibility(View.GONE);
                    digitalSignPath = "";
                    digitalSignBase64 = "";
                });
                LocalBroadcastManager.getInstance(this).registerReceiver(mServiceReceiver,
                        new IntentFilter("digital_sign"));


            }

            private void validateCylinder(String scannedBarcode) {
                if (scannedBarcode.isEmpty()) return;

                int idx = -1;
                for (int i = 0; i < cylinderList.size(); i++) {
                    if (scannedBarcode.equals(cylinderList.get(i).getBarcode_no())) {
                        idx = i;
                        break;
                    }
                }

                if (idx != -1) {
                    CylinderData c = cylinderList.get(idx);
                    if (!c.isValidated()) {
                        c.setValidated(true);
                        recyclerView.getAdapter().notifyItemChanged(idx);
                    }
                    Toast.makeText(this, "Cylinder " + scannedBarcode + " validated", Toast.LENGTH_SHORT).show();
                } else {
                    new AlertDialog.Builder(this)
                            .setTitle("त्रुटी")
                            .setMessage("स्कॅन केलेला बारकोड '" + scannedBarcode + "' यादीत आढळला नाही.")
                            .setPositiveButton("ठीक", (dialog, which) -> {
                                dialog.dismiss();
                            })
                            .setCancelable(false)
                            .show();
                }
            }

            private void postUsingVolley() {
                String dcno = getIntent().getStringExtra("dcno");
                String email = SharedPref.getInstance(this).getEmail();
                String vehicleNo = SharedPref.getInstance(this).getVehicleNo();
                String transType = "DEL";

                // Calculate validated quantity and cylinder numbers
                List<String> validatedCylinders = new ArrayList<>();
                for (CylinderData cylinder : cylinderList) {
                    if (cylinder.isValidated()) {
                        validatedCylinders.add(cylinder.getCyl_code());
                    }
                }
                String cylNos = String.join(",", validatedCylinders);
                String validatedQty = String.valueOf(validatedCylinders.size());

                // Validate inputs
                if (dcno == null || dcno.isEmpty() || validatedQty.equals("0") || cylNos.isEmpty() ||
                        email == null || email.isEmpty() || digitalSignBase64.isEmpty()) {
                    dialog.dismiss();
                    Toast.makeText(this, "Missing required data. Please ensure all fields are filled and signature is provided.", Toast.LENGTH_LONG).show();
                    return;
                }

                String dbHost = SharedPref.getInstance(this).getDBHost();
                String dbUsername = SharedPref.getInstance(this).getDBUsername();
                String dbPassword = SharedPref.getInstance(this).getDBPassword();
                String dbName = SharedPref.getInstance(this).getDBName();

                if (dbHost == null || dbUsername == null || dbPassword == null || dbName == null) {
                    dialog.dismiss();
                    Toast.makeText(this, "Database configuration missing.", Toast.LENGTH_LONG).show();
                    return;
                }

                String url = delivery_validation; // Replace with actual URL

                StringRequest stringRequest = new StringRequest(
                        Request.Method.POST, url,
                        response -> {
                            dialog.dismiss();
                            try {
                                ValidateDcResponse validateDcResponse = new Gson().fromJson(response, ValidateDcResponse.class);
                                if (validateDcResponse.getStatus() != null && !validateDcResponse.getStatus().isEmpty()) {
                                    if ("success".equals(validateDcResponse.getStatus())) {
                                        Toast.makeText(this, validateDcResponse.getMsg(), Toast.LENGTH_LONG).show();
                                        Intent i = new Intent(SecondValidateDcActivity.this, ViewDeliveryPrint.class);
                                        i.putExtra("no",validateDcResponse.getDcno());
                                        i.putExtra("type","DEL");
                                        startActivity(i);
                                        finish();
                                    } else {
                                        Toast.makeText(this, validateDcResponse.getMsg(), Toast.LENGTH_LONG).show();
                                    }
                                } else {
                                    Toast.makeText(this, "Invalid response from server", Toast.LENGTH_LONG).show();
                                }
                            } catch (Exception e) {
                                Log.e(TAG, "JSON parsing error: " + e.getMessage());
                                Toast.makeText(this, "Error parsing response", Toast.LENGTH_LONG).show();
                            }
                        },
                        error -> {
                            dialog.dismiss();
                            Log.e(TAG, "Volley error: " + (error.getMessage() != null ? error.getMessage() : "Unknown error"));
                            Toast.makeText(this, "Network error: " + (error.getMessage() != null ? error.getMessage() : "Unknown error"), Toast.LENGTH_LONG).show();
                        }) {
                    @Override
                    protected Map<String, String> getParams() throws AuthFailureError {
                        Map<String, String> params = new HashMap<>();
                        params.put("db_host", dbHost);
                        params.put("db_username", dbUsername);
                        params.put("db_password", dbPassword);
                        params.put("db_name", dbName);
                        params.put("dcno", dcno);
                        params.put("validated_qty", validatedQty);
                        params.put("cyl_nos", cylNos);
                        params.put("email", email);
                        params.put("trans_type", transType);
                        params.put("sign", digitalSignBase64);
                        params.put("lati", latitude);
                        params.put("logi", logitude);
                        params.put("vehicle_no", vehicleNo);
                        return params;
                    }
                };

                VolleySingleton.getInstance(this).addToRequestQueue(stringRequest);
            }

            private final BroadcastReceiver mServiceReceiver = new BroadcastReceiver() {
                @Override
                public void onReceive(Context context, Intent intent) {
                    if ("digital_sign".equals(intent.getAction())) {
                        String signed = intent.getStringExtra("Signed");
                        digitalSignPath = intent.getStringExtra("path") != null ? intent.getStringExtra("path") : "";
                        if ("true".equals(signed)) {
                            constraintSigned.setVisibility(View.VISIBLE);
                            File imgFile = new File(digitalSignPath);
                            if (imgFile.exists()) {
                                Bitmap myBitmap = BitmapFactory.decodeFile(imgFile.getAbsolutePath());
                                signedImg.setImageBitmap(myBitmap);
                                try {
                                    digitalSignBase64 = Base64.getEncoder().encodeToString(Files.readAllBytes(imgFile.toPath()));
                                } catch (IOException e) {
                                    Log.e(TAG, "Error encoding signature: " + e.getMessage());
                                    digitalSignBase64 = "";
                                }
                            }
                        }
                    }
                }
            };



            @Override
            public void onItemClick(int position) {
                // Not used, but required by OnItemClickListener
            }


    // dispatchKeyEvent removed. Logic moved to OnKeyListener.



    @Override
    public void locationOn() {
        Toast.makeText(this, "Location On", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void currentLocation(Location location) {
        latitude = String.valueOf(location.getLatitude());

        logitude = String.valueOf(location.getLongitude());

    }
    @Override
    protected void onResume() {
        super.onResume();
        easyWayLocation.startLocation();
    }

    @Override
    protected void onPause() {
        super.onPause();
        easyWayLocation.endUpdates();
    }

    @Override
    public void locationCancelled() {
        Toast.makeText(this, "Location Cancelled", Toast.LENGTH_SHORT).show();
    }

}

File: app/src/main/java/com/arnichem/arnichem_barcode/TransactionsView/validate_dc/SecondValidateDcActivity.kt
package com.arnichem.arnichem_barcode.TransactionsView.validate_dc

import android.app.ProgressDialog
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.ActivityInfo
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.arnichem.arnichem_barcode.Barcode.LaserScannerActivity
import com.arnichem.arnichem_barcode.Barcode.NewScanner
import com.arnichem.arnichem_barcode.R
import com.arnichem.arnichem_barcode.digital_signature.ActivityDigitalSignature
import com.arnichem.arnichem_barcode.util.SharedPref
import com.arnichem.arnichem_barcode.util.Util
import java.io.File
import java.lang.reflect.Type

class CylinderAdapter(
    private val context: Context,
    private val cylinderList: List<CylinderData>,
    private val onValidate: (String) -> Unit
) : RecyclerView.Adapter<CylinderAdapter.ViewHolder>() {

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val cylinderText: TextView = itemView.findViewById(android.R.id.text1)
        val tickImage: ImageView = itemView.findViewById(android.R.id.icon)
    }

    override fun onCreateViewHolder(parent: android.view.ViewGroup, viewType: Int): ViewHolder {
        val view = android.view.LayoutInflater.from(parent.context)
            .inflate(android.R.layout.simple_list_item_1, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val cylinder = cylinderList[position]
        holder.cylinderText.text = cylinder.cyl_code
        holder.tickImage.visibility = if (cylinder.isValidated) View.VISIBLE else View.GONE
        holder.itemView.setOnClickListener { onValidate(cylinder.cyl_code) }
    }

    override fun getItemCount() = cylinderList.size
}

class SecondValidateDcActivity : AppCompatActivity() {

    private lateinit var recyclerView: RecyclerView
    private lateinit var dcNoTextView: TextView
    private lateinit var invoiceNoTextView: TextView
    private lateinit var customerNameTextView: TextView
    private lateinit var submitButton: Button
    private lateinit var uploadSign: Button
    private lateinit var constraintSigned: ConstraintLayout
    private lateinit var signedImg: ImageView
    private lateinit var closeImg: ImageView
    private var dialog: ProgressDialog? = null
    private var cylinderList: List<CylinderData> = emptyList()
    private var digitalSignPath = ""
    private val TAG = "SecondValidateDcActivity"
    private var isAllFabsVisible = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_second_validate_dc)

        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.title = "Validate DC"

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT

        dcNoTextView = findViewById(R.id.dcNoTextView)
        invoiceNoTextView = findViewById(R.id.invoiceNoTextView)
        customerNameTextView = findViewById(R.id.customerNameTextView)
        recyclerView = findViewById(R.id.recyclerView)
        submitButton = findViewById(R.id.submitButton)
        uploadSign = findViewById(R.id.uploadSign)
        constraintSigned = findViewById(R.id.constraintSigned)
        signedImg = findViewById(R.id.signedImg)
        closeImg = findViewById(R.id.closeImg)

        val addFab = findViewById<com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton>(R.id.add_fab)
        val cameraScanFab = findViewById<com.google.android.material.floatingactionbutton.FloatingActionButton>(R.id.camera_scan)
        val barcodeScanFab = findViewById<com.google.android.material.floatingactionbutton.FloatingActionButton>(R.id.barcode_scan)

        cameraScanFab.visibility = View.GONE
        barcodeScanFab.visibility = View.GONE

        addFab.shrink()

        addFab.setOnClickListener {
            if (!isAllFabsVisible) {
                cameraScanFab.show()
                barcodeScanFab.show()
                addFab.extend()
                isAllFabsVisible = true
            } else {
                cameraScanFab.hide()
                barcodeScanFab.hide()
                addFab.shrink()
                isAllFabsVisible = false
            }
        }

        cameraScanFab.setOnClickListener {
            startActivity(Intent(this, NewScanner::class.java).apply { putExtra("type", "validate") })
        }

        barcodeScanFab.setOnClickListener {
            startActivity(Intent(this, LaserScannerActivity::class.java).apply { putExtra("type", "validate") })
        }

        val intent = intent
        val dcno = intent.getStringExtra("dcno")
        val invoiceNo = intent.getStringExtra("invoiceNo")
        val customerName = intent.getStringExtra("customerName")
        val cylinderNumbersJson = intent.getStringExtra("cylinderNumbers")

        val listType: Type = object : TypeToken<List<CylinderData>>() {}.type
        cylinderList = Gson().fromJson(cylinderNumbersJson, listType) ?: emptyList()

        dcNoTextView.text = "Dc No: $dcno"
        invoiceNoTextView.text = "Invoice No: $invoiceNo"
        customerNameTextView.text = "Customer Name: $customerName"

        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = CylinderAdapter(this, cylinderList) { scannedBarcode ->
            validateCylinder(scannedBarcode)
        }

        findViewById<TextView>(R.id.no_data).visibility = if (cylinderList.isEmpty()) View.VISIBLE else View.GONE

        submitButton.setOnClickListener {
            dialog = ProgressDialog(this).apply {
                setTitle("Submitting")
                setMessage("Please wait....")
                setProgressStyle(ProgressDialog.STYLE_SPINNER)
                show()
            }
            // Simulate submission (replace with actual API call)
            dialog?.dismiss()
            Toast.makeText(this, "Validation submitted", Toast.LENGTH_SHORT).show()
        }

        uploadSign.setOnClickListener {
            startActivity(Intent(this, ActivityDigitalSignature::class.java).apply { putExtra("type", "validate") })
        }

        closeImg.setOnClickListener {
            SharedPref.getInstance(this).setSign("")
            constraintSigned.visibility = View.GONE
        }
    }

    private fun validateCylinder(scannedBarcode: String) {
        cylinderList.forEach { cylinder ->
            if (cylinder.cyl_code == scannedBarcode || cylinder.barcode_no == scannedBarcode) {
                cylinder.isValidated = true
                recyclerView.adapter?.notifyDataSetChanged()
                Toast.makeText(this, "Cylinder $scannedBarcode validated", Toast.LENGTH_SHORT).show()
                return
            }
        }
        Toast.makeText(this, "No matching cylinder found for $scannedBarcode", Toast.LENGTH_SHORT).show()
    }

    override fun onResume() {
        super.onResume()
        LocalBroadcastManager.getInstance(this).registerReceiver(mServiceReceiver, IntentFilter("digital_sign"))
    }

    override fun onPause() {
        super.onPause()
        LocalBroadcastManager.getInstance(this).unregisterReceiver(mServiceReceiver)
    }

    private val mServiceReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (intent.action == "digital_sign") {
                val signed = intent.getStringExtra("Signed")
                digitalSignPath = intent.getStringExtra("path") ?: ""
                if (signed == "true") {
                    constraintSigned.visibility = View.VISIBLE
                    val imgFile = File(digitalSignPath)
                    if (imgFile.exists()) {
                        val myBitmap = android.graphics.BitmapFactory.decodeFile(imgFile.absolutePath)
                        signedImg.setImageBitmap(myBitmap)
                    }
                }
            }
        }
    }
}